File: scripts/compress-images.mjs
/**
 * compress-images.mjs
 * One-off script — compresses every large JPEG/PNG in public/assets/
 * that is used as a photo (skips logos, placeholder, hero images that
 * are already optimised by next/image at runtime).
 *
 * Run:  node scripts/compress-images.mjs
 *
 * Uses `sharp` installed as a devDependency just for this script.
 * Output: overwrites files in-place; originals are backed up to
 *         public/assets/_originals/ (created automatically).
 */

import sharp from 'sharp';
import { readdir, mkdir, copyFile, stat } from 'fs/promises';
import { join, extname, basename } from 'path';
import { existsSync } from 'fs';

const ASSETS = new URL('../public/assets', import.meta.url).pathname.replace(/^\/([A-Z]:)/, '$1');
const BACKUP = join(ASSETS, '_originals');

// Files to SKIP (logos, placeholder — already small or need transparency kept)
const SKIP = new Set([
  'ATA-logo.png', 'Black-Canvas-logo.png', 'CEA-logo.png', 'JKOA-logo.png',
  'KVK-logo.jpeg', 'apiit-logo.png', 'eclub-logo.png', 'hi-tech-logo.png',
  'life-vision-logo.jpeg', 'sab-logo.png', 'sampath-logo.jpeg',
  'sasnaka-logo.png', 'sandbox-logo.png', 'unilever-logo.png',
  'veerakesari-logo.png', 'placeholder-image.png',
]);

const IMAGE_EXTS = new Set(['.jpg', '.jpeg', '.png', '.webp']);

// Target: resize so longest edge ≤ 1600px, JPEG quality 78, PNG effort 9
const MAX_EDGE = 1600;
const JPEG_QUALITY = 78;
const WEBP_QUALITY = 80;
const PNG_COMPRESSION = 9;

async function compress(filePath, name) {
  const ext = extname(name).toLowerCase();
  const img = sharp(filePath);
  const meta = await img.metadata();
  const { width, height } = meta;

  // Determine output dimensions
  let newW = width, newH = height;
  if (width > MAX_EDGE || height > MAX_EDGE) {
    if (width >= height) { newW = MAX_EDGE; newH = Math.round(height * MAX_EDGE / width); }
    else                  { newH = MAX_EDGE; newW = Math.round(width  * MAX_EDGE / height); }
  }

  const resized = img.resize(newW, newH, { withoutEnlargement: true });

  if (ext === '.jpg' || ext === '.jpeg') {
    await resized.jpeg({ quality: JPEG_QUALITY, mozjpeg: true }).toFile(filePath + '.tmp');
  } else if (ext === '.png') {
    await resized.png({ compressionLevel: PNG_COMPRESSION }).toFile(filePath + '.tmp');
  } else if (ext === '.webp') {
    await resized.webp({ quality: WEBP_QUALITY }).toFile(filePath + '.tmp');
  }

  // Swap tmp → original (sharp can't overwrite in-place)
  const { rename } = await import('fs/promises');
  await rename(filePath + '.tmp', filePath);
}

async function run() {
  if (!existsSync(BACKUP)) await mkdir(BACKUP, { recursive: true });

  const files = await readdir(ASSETS);
  const targets = files.filter(f => IMAGE_EXTS.has(extname(f).toLowerCase()) && !SKIP.has(f) && !f.startsWith('_'));

  console.log(`\n🗜  Compressing ${targets.length} images (target: ≤${MAX_EDGE}px longest edge, JPEG q${JPEG_QUALITY})\n`);

  let saved = 0;
  for (const name of targets) {
    const filePath = join(ASSETS, name);
    const backupPath = join(BACKUP, name);

    const before = (await stat(filePath)).size;

    // Back up original (skip if already backed up)
    if (!existsSync(backupPath)) await copyFile(filePath, backupPath);

    try {
      await compress(filePath, name);
      const after = (await stat(filePath)).size;
      const pct = Math.round((1 - after / before) * 100);
      saved += (before - after);
      console.log(`  ✓ ${name.padEnd(55)} ${(before/1024).toFixed(0).padStart(6)}KB → ${(after/1024).toFixed(0).padStart(5)}KB  (-${pct}%)`);
    } catch (err) {
      console.error(`  ✗ ${name}: ${err.message}`);
    }
  }

  console.log(`\n✅ Done. Total saved: ${(saved / 1024 / 1024).toFixed(1)} MB`);
  console.log(`   Originals backed up to: public/assets/_originals/\n`);
}

run().catch(console.error);

File: scripts/compress_images.py
"""
compress_images.py
Compresses all large JPEG/PNG photos in public/assets/ using Pillow.

- Resizes so longest edge <= 1600px (preserves aspect ratio)
- JPEG quality 78 (good visual quality, ~60-80% file size reduction)
- Backs up originals to public/assets/_originals/ first
- Skips logos, placeholder, and already-small files (< 300 KB)

Run: python scripts/compress_images.py
"""

import os
import shutil
from pathlib import Path
from PIL import Image

ASSETS = Path(__file__).parent.parent / "public" / "assets"
BACKUP = ASSETS / "_originals"

MAX_EDGE = 1600
JPEG_QUALITY = 78
SKIP_UNDER_BYTES = 300 * 1024  # 300 KB — already web-ready

# Files to skip — logos (may need transparency), placeholder, hero images
SKIP_NAMES = {
    "ATA-logo.png", "Black-Canvas-logo.png", "CEA-logo.png", "JKOA-logo.png",
    "KVK-logo.jpeg", "apiit-logo.png", "eclub-logo.png", "hi-tech-logo.png",
    "life-vision-logo.jpeg", "sab-logo.png", "sampath-logo.jpeg",
    "sasnaka-logo.png", "sandbox-logo.png", "unilever-logo.png",
    "veerakesari-logo.png", "placeholder-image.png",
}

IMAGE_EXTS = {".jpg", ".jpeg", ".png", ".webp"}

def compress(src: Path, backup: Path):
    """Resize + re-encode one image file in-place."""
    if not backup.exists():
        shutil.copy2(src, backup)

    img = Image.open(src)

    # Convert RGBA/P to RGB for JPEG (can't save transparency as JPEG)
    if src.suffix.lower() in (".jpg", ".jpeg") and img.mode in ("RGBA", "P", "LA"):
        img = img.convert("RGB")

    w, h = img.size
    if w > MAX_EDGE or h > MAX_EDGE:
        ratio = MAX_EDGE / max(w, h)
        img = img.resize((int(w * ratio), int(h * ratio)), Image.LANCZOS)

    if src.suffix.lower() in (".jpg", ".jpeg"):
        img.save(src, "JPEG", quality=JPEG_QUALITY, optimize=True, progressive=True)
    elif src.suffix.lower() == ".png":
        img.save(src, "PNG", optimize=True)
    elif src.suffix.lower() == ".webp":
        img.save(src, "WEBP", quality=JPEG_QUALITY, method=6)

def main():
    BACKUP.mkdir(exist_ok=True)

    targets = [
        f for f in ASSETS.iterdir()
        if f.is_file()
        and f.suffix.lower() in IMAGE_EXTS
        and f.name not in SKIP_NAMES
        and not f.name.startswith("_")
    ]
    targets.sort(key=lambda f: f.name)

    print(f"\n🗜  {len(targets)} images to process (target: ≤{MAX_EDGE}px, JPEG q{JPEG_QUALITY})\n")

    total_saved = 0
    processed = 0
    skipped = 0

    for f in targets:
        before = f.stat().st_size
        if before < SKIP_UNDER_BYTES:
            print(f"  ⏭  {f.name:<60} {before/1024:.0f} KB — already small, skipped")
            skipped += 1
            continue

        backup_path = BACKUP / f.name
        try:
            compress(f, backup_path)
            after = f.stat().st_size
            pct = round((1 - after / before) * 100)
            total_saved += before - after
            processed += 1
            print(f"  ✓  {f.name:<60} {before/1024/1024:.1f} MB → {after/1024:.0f} KB  (-{pct}%)")
        except Exception as e:
            print(f"  ✗  {f.name}: {e}")

    print(f"\n{'─'*70}")
    print(f"✅  Done!")
    print(f"   Processed : {processed} files")
    print(f"   Skipped   : {skipped} files (already ≤300 KB)")
    print(f"   Total saved: {total_saved / 1024 / 1024:.1f} MB")
    print(f"   Originals backed up → public/assets/_originals/\n")

if __name__ == "__main__":
    main()
